runtimes/common/src/migrations.rs

```diff
@@ -16,67 +16,77 @@
+		assert!(MigrationCounter::<T>::get().is_zero());
 
-		match (has_migration_storage, has_filter_storage) {
-			(false, false) => Err("Pallets not present"),
-			(true, false) => Err("DynFilter not present"),
-			(false, true) => Err("RelayMigration not present"),
-			_ => Ok(()),
-		}
+		// Use iter_keys() on new storage so it won't try to decode values.
+		let ctypes_to_migrate = ctype::Ctypes::<T>::iter_keys().count();
+
+		log::info!("🪪  CType pallet pre check: {:?} CTypes to migrate", ctypes_to_migrate);
+
+		MigrationCounter::<T>::set(ctypes_to_migrate as u32);
+		Ok(())
 	}
 
 	fn on_runtime_upgrade() -> frame_support::weights::Weight {
-		let entries: u32 = 2;
-		if frame_support::storage::unhashed::clear_prefix(&Twox128::hash(b"RelayMigration"), Some(entries), None)
-			.maybe_cursor
-			.is_some()
-		{
-			log::warn!("Pallet RelayMigration not removed entirely")
-		}
-		if frame_support::storage::unhashed::clear_prefix(&Twox128::hash(b"DynFilter"), Some(entries), None)
-			.maybe_cursor
-			.is_some()
-		{
-			log::warn!("Pallet DynFilter not removed entirely")
-		}
+		let current = ctype::Pallet::<T>::current_storage_version();
+		let onchain = ctype::Pallet::<T>::on_chain_storage_version();
+
+		log::info!(
+			"💰 Running migration with current storage version {:?} / onchain {:?}",
+			current,
+			onchain
+		);
+
+		let mut num_translations = 0u64;
+		let default_block_number = <T as frame_system::Config>::BlockNumber::zero();
 
-		<R as frame_system::Config>::DbWeight::get().writes((entries * 2).into())
+		ctype::Ctypes::<T>::translate_values(|old: CtypeCreatorOf<T>| {
+			num_translations = num_translations.saturating_add(1);
+			Some(CtypeEntryOf::<T> {
+				creator: old,
+				creation_block_number: default_block_number,
+			})
+		});
+		current.put::<ctype::Pallet<T>>();
+
+		// Num translations + old version read and new version write
+		T::DbWeight::get().reads_writes(num_translations.saturating_add(1), num_translations.saturating_add(1))
 	}
 
 	#[cfg(feature = "try-runtime")]
 	fn post_upgrade() -> Result<(), &'static str> {
-		log::info!("Post check RemoveRelocationPallets.");
-		let has_migration_storage = frame_support::storage::migration::have_storage_value(
-			b"RelayMigration",
-			b"RelayNumberStrictlyIncreases",
-			b"",
-		);
-		let has_filter_storage = frame_support::storage::migration::have_storage_value(b"DynFilter", b"Filter", b"");
+		assert_eq!(ctype::Pallet::<T>::on_chain_storage_version(), 1);
 
-		match (has_migration_storage, has_filter_storage) {
-			(false, false) => Ok(()),
-			(true, false) => Err("RelayMigration still present"),
-			(false, true) => Err("DynFilter still present"),
-			(true, true) => Err("Pallets still present"),
+		// Use iter() on new storage so it also checks that the new values can be
+		// decoded after the migration.
+		assert_eq!(MigrationCounter::<T>::get(), ctype::Ctypes::<T>::iter().count() as u32);
+		if let Some(ctype_entry) = ctype::Ctypes::<T>::iter_values().last() {
```

Comment:
To be 100% safe, we should check every element and not only the last one.

Reply:
https://github.com/KILTprotocol/kilt-node/pull/440/commits/8b4c9dcc68390383f61372d428cdcbbddf6bcdcc